write-fonts/src/tables/cmap.rs

```diff
@@ -2,62 +2,142 @@
+    + 2 * seg_count * 4  // 4 parallel arrays of len seg_count, 2 bytes per entry
+    + 2 * gid_count // 2 bytes per gid in glyphIdArray
+}
+
+impl CmapSubtable {
+    fn create_format_4(
+        lang: u16,
+        end_code: Vec<u16>,
+        start_code: Vec<u16>,
+        id_deltas: Vec<i16>,
+    ) -> Self {
+        if HashSet::from([start_code.len(), end_code.len(), id_deltas.len()]).len() != 1 {
+            panic!("uneven parallel arrays, very bad. Very very bad.");
+        }
+
+        let seg_count: u16 = start_code.len().try_into().unwrap();
+        // Spec: Log2 of the maximum power of 2 less than or equal to segCount (log2(searchRange/2),
+        // which is equal to floor(log2(segCount)))
+        let entry_selector = (seg_count as f32).log2().floor();
+
+        // Spec: Maximum power of 2 less than or equal to segCount, times 2
+        // ((2**floor(log2(segCount))) * 2, where “**” is an exponentiation operator)
+        let search_range = 2u16.pow(entry_selector as u32).checked_mul(2).unwrap();
+
+        // if 2^entry_selector*2 is a u16 then so is entry_selector
+        let entry_selector: u16 = entry_selector as u16;
+        let range_shift: u16 = seg_count * 2 - search_range;
+
+        let id_range_offsets = id_deltas.iter().map(|_| 0).collect();
+        CmapSubtable::Format4(Cmap4::new(
+            size_of_cmap4(seg_count, 0),
+            lang,
+            seg_count * 2,
+            search_range,
+            entry_selector,
+            range_shift,
+            end_code,
+            start_code,
+            id_deltas,
+            id_range_offsets,
+            vec![], // becauseour idRangeOffset's are 0 glyphIdArray is unused
+        ))
+    }
+}
+
+impl Cmap {
+    /// Generates a [cmap](https://learn.microsoft.com/en-us/typography/opentype/spec/cmap) that is expected to work in most modern environments.
+    ///
+    /// For the time being just emits [format 4](https://learn.microsoft.com/en-us/typography/opentype/spec/cmap#format-4-segment-mapping-to-delta-values)
+    /// so we can drive towards compiling working fonts. In time we may wish to additionally emit format 12 to support
+    /// novel codepoints.
+    pub fn from_mappings(mappings: impl IntoIterator<Item = (char, GlyphId)>) -> Cmap {
+        let mut end_code: Vec<u16> = Vec::new();
+        let mut start_code: Vec<u16> = Vec::new();
+        let mut id_deltas: Vec<i16> = Vec::new();
+        let mut id_range_offsets: Vec<u16> = Vec::new();
+
+        let mut prev = (u16::MAX - 1, u16::MAX - 1);
+        for (cp, gid) in mappings.into_iter() {
+            let cp = (cp as u32).try_into().unwrap();
+            let next_in_run = (
+                prev.0.checked_add(1).unwrap(),
+                prev.1.checked_add(1).unwrap(),
+            );
+            let current = (cp, gid.to_u16());
+            // Codepoint and gid need to be continuous
+            if current != next_in_run {
+                // Start a new run
```

Comment:
fonttools's `_c_m_a_p` module contain a `splitRange` function that attempts 

> to split a range of character codes into subranges with consecutive glyph IDs in such a way that the cmap4 subtable can be stored "most" efficiently

https://github.com/fonttools/fonttools/blob/444a349ef4dd1bae2d43f6ea0e06d73fb6466b9a/Lib/fontTools/ttLib/tables/_c_m_a_p.py#L783

I wonder if we should do something like that here as well

Reply:
I think we should but perhaps as a followon

Reply:
Deferred to https://github.com/googlefonts/fontations/issues/282